### ISSUE TITLE
Inconsistent behaviour of plugin enqueue method when inputs has empty shapes(i.e. 0 on batch dimension)

### ISSUE THREAD

**Issue by kfiring** [NVIDIA/TensorRT#2833]
## Description

i got a onnx model(maskrcnn) and converted it into trt engine by using trtexec, there are 2 custom ops in onnx model, so i implemented 2 corresponding plugins in tensorrt, one is NMSRotatedPlugin and another is ROIAlignRotatedPlugin. when testing inference on the converted trt engine, i found that when inputs has empty shapes, the inputs/outputs parameters of enqueue method of these 2 plugins are given differently by tensorrt. 

As for ROIAlignRotatedPlugin, it has 2 inputs and 1 output, when the 1st dimension of the 2nd input is 0(which means no rois), the output should be empty(0 in the 1st dim). **the inputs/outputs parameter are non-null pointer** and the "checked the 0 condition and return" logic works fine, as show in the following picture 
![image](https://user-images.githubusercontent.com/20376105/228807907-a25a0a2b-d611-4b93-a36b-397f72331d82.png)

As for NMSRotatedPlugin, it has 3 inputs and 2 outputs, when the 1st dim of the first  2 inputs are 0(which means no bbox were found), the 1st output should also be empty(0 in the 1st dim). but this time **the inputs/outputs parameter are given null pointers**, and **after "checked the 0 condition and return" logic, tensorrt throws out  "illegal memory access" error**, as show in followig pictures:
![image](https://user-images.githubusercontent.com/20376105/228808186-381a824c-6ccf-4c36-ba3e-2b9c3d69496f.png)

I think the illegal memory access is related to the null-pointers and i don't know why the parameter passing behaviour are different(**i.e. one gets non-null pointers and the other get null**) on these 2 plugins.

## Environment

**TensorRT Version**:  8.6.0
**GPU Type**: RTX 3090
**Nvidia Driver Version**: 470.129.06
**CUDA Version**: 11.4
**CUDNN Version**: 8.8.0
**Operating System + Version**: Ubuntu 20.04.4 LTS
**Python Version (if applicable)**: 3.8
**TensorFlow Version (if applicable)**: none
**PyTorch Version (if applicable)**: 1.9
**Baremetal or Container (if container which image + tag)**: container

**Comment by zerollzeng** [comment]
@samurdhikaru Can you help? Thanks!

**Comment by kfiring** [comment]
is anyone take care of this issue?this issue has block my work for days...

**Comment by samurdhikaru** [comment]
@kfiring For empty I/O tensors, TRT has so far not made any guarantees as to the nature of the corresponding buffer pointers -- the only guarantee is that for non-empty I/O tensors, a pre-allocated device buffer of the appropriate size will be passed into `enqueue`. So it is the plugin author's responsibility to inspect the `PluginTensorDesc` of each I/O tensor to check whether that tensor is empty or not (if it is empty, you should not be attempting to read from/write to it).

**Comment by kfiring** [comment]
> 

of course i checked the `PluginTensorDesc`, when i found it's empty, i returned immediately, as show in below:
![image](https://user-images.githubusercontent.com/20376105/230526129-1bb21b3b-9e39-4cb9-bc8d-c07b4578bbd9.png)
now the question is why i got this error: `[E] 1: [deviceToShapeHostRunner.cpp::execute::38] Error Code 1: Cuda Runtime (an illegal memory access was encountered)` after NMSRotatedPlugin got empty tensors.

**Comment by kfiring** [comment]
> @kfiring For empty I/O tensors, TRT has so far not made any guarantees as to the nature of the corresponding buffer pointers -- the only guarantee is that for non-empty I/O tensors, a pre-allocated device buffer of the appropriate size will be passed into `enqueue`. So it is the plugin author's responsibility to inspect the `PluginTensorDesc` of each I/O tensor to check whether that tensor is empty or not (if it is empty, you should not be attempting to read from/write to it).

of course i checked the `PluginTensorDesc`, when i found it's empty, i returned immediately, as show in below:
![image](https://user-images.githubusercontent.com/20376105/230526129-1bb21b3b-9e39-4cb9-bc8d-c07b4578bbd9.png)
now the question is why i got this error: `[E] 1: [deviceToShapeHostRunner.cpp::execute::38] Error Code 1: Cuda Runtime (an illegal memory access was encountered)` after NMSRotatedPlugin got empty tensors.

**Comment by samurdhikaru** [comment]
@kfiring Can you please provide a minimal repro for the issue with the `NMSRotatedPlugin`?

**Comment by kfiring** [comment]
> @kfiring Can you please provide a minimal repro for the issue with the `NMSRotatedPlugin`?

@samurdhikaru i uploaded codes of NMSRotatedPlugin [here](https://drive.google.com/drive/folders/1xxK64BseyBzUzi3lb4Kj8ZDYSMfX0J6b?usp=share_link) and ROIAlignRotatedPlugin [here](https://drive.google.com/drive/folders/1tQeIIh_auOLnEg_GjWWkjR-2kTKvwacs?usp=share_link), please take a look.

**Comment by kfiring** [comment]
> @kfiring Can you please provide a minimal repro for the issue with the `NMSRotatedPlugin`?

hi, @samurdhikaru, i uploaded all stuff [here](https://drive.google.com/drive/folders/1VENdj7t28dg3oQnHOATzxvGRjG69Q9-g?usp=share_link) you may need to reproduce the error. 
reproduce steps:
1. put `nmsRotatedPlugin` and `roiAlignRotatedPlugin` into `<path_to_tensorrt_source>/plugin`.
2. put `mixedai.h` into `<path_to_tensorrt_source>/plugin/common`.
3. replace `<path_to_tensorrt_source>/parsers/onnx/builtin_op_importers.cpp` with mine.
4. modify `<path_to_tensorrt_source>/plugin/CMakeLists.txt` to add `nmsRotatedPlugin` and `roiAlignRotatedPlugin` into compile process.
5. recompile tensorrt.
6. then you can convert the [udated_modle.onnx](https://drive.google.com/file/d/1QZhJwxEk9dbZB2dWhvXWSROCyfznDov1/view?usp=share_link) into tensorrt engine file by using command `trtexec --onnx=updated_model.onnx --saveEngine=updated_model.trt --minShapes=image:0x240x320x3 --optShapes=image:1x480x640x3 --maxShapes=image:1x480x640x3 --verbose` or you can simply use my converted tensorrt engine file [updated_model.trt](https://drive.google.com/file/d/1AVzb2JYxRckTDpmhrcSIHTU_-hk7N6U3/view?usp=share_link)
7. now you can use my testing python script [modeltest_for_nvidia.py](https://drive.google.com/file/d/1NvxUJB46_CUhNbyEdwHjmfzSkNr651Ak/view?usp=share_link) to reproduce the error.

**Comment by kfiring** [comment]
another 10 days passed with no reply, is there anybody taking care of this?
